sw/airborne/modules/computer_vision/snake_gate_detection.c

```diff
@@ -0,0 +1,981 @@
+ * @param[in] quality The ratio of rightly colored pixels, 1.0 is best, 0.0 is worst.
+ *            If too many pixels inside the gate are of the right color, the quality is also 0.
+ *  @param[in] sides How many of the sides are sufficiently colored?
+ */
+extern void check_gate_initial(struct image_t *im, struct gate_img gate, float *quality, int *n_sides)
+{
+  int n_points, n_colored_points;
+  n_points = 0;
+  n_colored_points = 0;
+  int np, nc;
+
+  // how much of the side should be visible to count as a detected side?
+  float min_ratio_side = 0.30;
+  (*n_sides) = 0;
+
+  // check the four lines of which the gate consists:
+  struct point_t from, to;
+
+  from.x = gate.x - gate.sz;
+  from.y = gate.y - gate.sz_left;
+  to.x = gate.x - gate.sz;
+  to.y = gate.y + gate.sz_left;
+  check_line(im, from, to, &np, &nc);
+  if ((float) nc / (float) np >= min_ratio_side) {
+    (*n_sides)++;
+  }
+  n_points += np;
+  n_colored_points += nc;
+
+  from.x = gate.x - gate.sz;
+  from.y = gate.y + gate.sz_left;
+  to.x = gate.x + gate.sz;
+  to.y = gate.y + gate.sz_right;
+  check_line(im, from, to, &np, &nc);
+  if ((float) nc / (float) np >= min_ratio_side) {
+    (*n_sides)++;
+  }
+  n_points += np;
+  n_colored_points += nc;
+
+  from.x = gate.x + gate.sz;
+  from.y = gate.y + gate.sz_right;
+  to.x = gate.x + gate.sz;
+  to.y = gate.y - gate.sz_right;
+  check_line(im, from, to, &np, &nc);
+  if ((float) nc / (float) np >= min_ratio_side) {
+    (*n_sides)++;
+  }
+  n_points += np;
+  n_colored_points += nc;
+
+  from.x = gate.x + gate.sz;
+  from.y = gate.y - gate.sz_right;
+  to.x = gate.x - gate.sz;
+  to.y = gate.y - gate.sz_left;
+  check_line(im, from, to, &np, &nc);
+  if ((float) nc / (float) np >= min_ratio_side) {
+    (*n_sides)++;
+  }
+
+  n_points += np;
+  n_colored_points += nc;
+
+
+  // the quality is the ratio of colored points / number of points:
+  if (n_points == 0) {
+    (*quality) = 0;
+  } else {
+    (*quality) = ((float) n_colored_points) / ((float) n_points);
+  }
+
+  // check that the inside of the gate is not of the target color as well:
+  int n_samples_in = 20;
+  float num_color_center = 0;
+  float center_discard_threshold = 0.25;
+  for (int i = 0; i < n_samples_in; i++) {
+    // get a random coordinate:
+    int x_in = gate.x + (rand() % gate.sz) - (0.5 * gate.sz);
+    int y_in = gate.y + (rand() % gate.sz) - (0.5 * gate.sz);
+
+    // check if it has the right color
+    if (check_color_sgd(im, x_in, y_in)) {
+      num_color_center ++;
+    }
+  }
+  //how much center pixels colored?
+  float center_factor = num_color_center / (float)n_samples_in;
+  if (center_factor > center_discard_threshold) { (*quality) = 0; }
+
+}
+
+/**
+ * Determine the segment length between two 2D-points.
+ *
+ * @param[in] Q1 Point 1.
+ * @param[in] Q2 Point 2.
+ */
+float segment_length(struct point_t Q1, struct point_t Q2)
+{
+
+  float r = sqrt((Q1.x - Q2.x) * (Q1.x - Q2.x) + (Q1.y - Q2.y) * (Q1.y - Q2.y));
```

Comment:
ouch. Was this simply not working then? I changed it.

Reply:
No, it would still work, internally, the values will just be converted to doubles then run the computation using doubles then convert it back to float. Calling the float function directly is just for efficiency.